### ISSUE TITLE
Reference to freed object can return a random different one

### ISSUE THREAD

**Issue by jaykyburz** [godotengine/godot#32383]
**Godot version & OS/device including version:**
3.1.1 Windows 10. 

**Issue description:**
I think this is a simple explanation of an issue I have on my project. I have raised this issue without an example project because it will take some time to reproduce in a simple form and I would like some godot experts to look at what I am describing, and let me know if the issue is already logged and known (and perhaps fixed) before I spend more time working on a reproduction cases

----
Say I have a RobotFactory class and it can instantiate Robots, but it is only allowed to have one active in the world at once.  It saves the instantiated robot in a var. 

var my_robot = RobotScene.instance()

The robot walks around the world for a while an then is destroyed by something so in its own script it calls queue_free() on itself. 

Sometime later in the RobotFactory we call is_instance_valid(my_robot) to see if we should create a new robot. 

In my project I have found that sometimes is_instance_valid(my_robot) returns true after my_robot has been freed, but that it is pointing to some other random object in the scene that has been instantiated. Perhaps sometime after the queue_free, but before is_instance_valid call. 

The only thing I think could be happening is that there is some kind of underlying pooling system and my_robot is pointing to somewhere that is being reused after the robot is freed. 

Please let me know if this is a known issue, is already logged, or if I should do some more work to explain the issue better. 

---

*Bugsquad edit:* Keyword: previously freed instance

**Comment by KoBeWi** [comment]
Oooh, it had totally the same thing. Good to know I'm not alone ;_;

In my case, I had a Control node with script that calls `queue_free()` inside `_ready()` on some condition. Then another node referenced it in an `onready` variable. Later in code I had something like:
```
if is_instance_valid(my_control):
    my_control.do_stuff()
```
And it raised an error ~ "No method do_stuff in class ParticleMaterial". Like, what XD
It just sometimes takes random object or resource from the scene. But now always, it depends on when I use it. I tried to make reproduction project, but I couldn't >_> Although I'm instancing lots of objects in my scene using `ResourceInteractiveLoader` and I didn't test it with minimal project yet.

Ultimately I ended with this ;_;
```
if is_instance_valid(my_control) and my_control is Control:
    my_control.do_stuff()
```

**Comment by Zylann** [comment]
Could it be that the object ID got re-used? Also, are there threads involved? (I mean, anywhere, not just scripts).

**Comment by bojidar-bg** [comment]
Can confirm on a game I made some time ago without threads.
In it, I had enemies with the following snipped of code:
```gdscript
func ....():
        if target == null or not is_instance_valid(target):
                var players = get_tree().get_nodes_in_group("player")
                if players.size() == 0: return
                target = players[randi() % players.size()]
```
It worked -- enemies targeted random players, and switched when their target died.
But, when all players died, and a new enemy spawned, all other enemies targeted that enemy, due to the reference being reused.

**Comment by jaykyburz** [comment]
Attached is a sample project the demonstrates the issue.  It's basically what I described above. 

One factory produces red cubes and keeps a reference to the cube it last created.  It points to its reference using the pink squashed sphere in the screen shot below.  The red factly produces cubes every 4 seconds but they are freed after 1 second so there is 3 seconds where the is_instance_valid(my_instance) should return false. 

There is a second factory producing blue cubes every 1 second. 



![image](https://user-images.githubusercontent.com/1171070/65805091-a2f1ae00-e1c7-11e9-85fc-d79975e877c1.png)
We find that every so often, is_instance_valid(my_instance) on the red factory returns true when it should not, but is pointing to one of the blue cubes.
![image](https://user-images.githubusercontent.com/1171070/65805109-adac4300-e1c7-11e9-829d-163865767ad2.png)

[IsInstanceValid.zip](https://github.com/godotengine/godot/files/3664683/IsInstanceValid.zip)

**Comment by reduz** [comment]
This is a known problem, but it requires some major changes to fix. 4.0 will change the way objects are referenced so this (and other similar issues) can be properly fixed.

**Comment by jaykyburz** [comment]
thanks for the update

**Comment by vnen** [comment]
Pretty sure there's already a report about this.

**Comment by KoBeWi** [comment]
I found a workaround for some cases. Let's say you have an object stored in a variable called `my_instance`. You can just do
```
my_instance.connect("tree_exited", self, "set", ["my_instance", null])
```
Then instead of using `is_instance_valid()` you can just do `if my_instance`, which will now reliably point to null if object is freed. This of course breaks if you use `remove_child`.

**Comment by jaykyburz** [comment]
I wrote a little helper class and now try never to save a reference to a node directly.  

`extends Node
class_name NodeRefrence

var __ref:Node

func get_ref()->Node:
	return __ref

func set_ref(ref:Node):
	clear_ref()
	__ref = ref
	__ref.connect("tree_exiting", self, "ref_tree_exiting")

func clear_ref():
	if __ref:
		__ref.disconnect("tree_exiting", self, "ref_tree_exiting")
	__ref = null

func ref_tree_exiting():
	__ref = null`

then to use this you have some code that looks something like this. 

`
var target_ref = NodeRefrence.new()

func _ready():
    target_ref.set_ref(some_node)
   ....
    var target = target_ref.get_ref()


`

**Comment by akien-mga** [comment]
See #22395 for other example projects that reproduce it.

It sucks that we can't have a proper fix for this before 4.0 :/

**Comment by lawnjelly** [comment]
When I tried jaykyburz's test project I found that using WeakRef instead of a raw object pointer also fixed it for me.

i.e.
```
var my_robot_ref : WeakRef

my_robot_ref = weakref(bot)

	if my_robot_ref:
		var ref = my_robot_ref.get_ref()
		if ref:
			$RedPointer.global_transform = ref.global_transform
```
WeakRef internally uses an ObjectID rather than pointer, which is looked up when you call get_ref().

If you store both the original pointer and a weakref in the test project, if you remote debug it, you can clearly see the point at which the pointer starts pointing to the wrong object.

**Comment by ttencate** [comment]
@akien-mga 
> It sucks that we can't have a proper fix for this before 4.0 :/

Agreed. Maybe it would be good to add some warnings to the 3.x docs? I would expect to find such warnings in these places:

* https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_basics.html#memory-management
* https://docs.godotengine.org/en/stable/classes/class_@gdscript.html#class-gdscript-method-is-instance-valid
* https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-free

**Comment by Yekez** [comment]
I had an ultimate case in my game. I was not aware of such a thing before reading about this dangling var issue. My script was "queue_free()"ing multiple objects and reasigning new objects to those same variables. Took me pretty long time to find out what was wrong. Documenting this somewhere, if not fixing it, might help people a lot.

**Comment by remram44** [comment]
I'm running into this again and wondering if the docs shouldn't be improved. @lawnjelly mentions that weakref works in this case (https://github.com/godotengine/godot/issues/32383#issuecomment-576401878) but the doc for `WeakRef` only says:

> If this object is not a reference, weakref still works, however, it does not have any effect on the object.

It is not clear what "still works" means. Does `get_ref()` correctly returns `null` if the object is gone, even if it has been re-allocated?

**Comment by lawnjelly** [comment]
It looks like it returns either a reference to the object, or a NULL variant.

If that theory is correct, then there may be a lot of tutorials using straight pointers that should probably be using WeakRefs. For safety, a pointer to any other object that is not local, should probably be a WeakRef.

It may even be possible to rig GDScript to do this with no change by the user. I.e. member variables use WeakRefs and only local variables use straight pointers (if even local variables should be considered 'safe'). It depends how quick the lookup by object ID is I guess. But the present situation is inherently unsafe I think.

Usually in a scripting language you will use some kind of handle for this, and disallow raw pointers. And maybe that is what WeakRef effectively does, but it is just not well known. And maybe this is fixed in Godot 4, but still remains in 3.x.

**Comment by Zylann** [comment]
> maybe a lot of tutorials using straight pointers should be using WeakRefs.

In GDScript at least, weakref is not needed (and even kinda degenerate) when the referenced object is an Object (and not a Reference). Either way, `is_instance_valid()` is the function to use to check if the pointer is valid or not.
The intented use of WeakRef is to not contribute to Reference's refcount, which it achieves by using ObjectID. The fact it allowes checking for validity of the pointer is a side-effect that isnt much different from `is_instance_valid()`.

**Comment by lawnjelly** [comment]
> Either way, `is_instance_valid()` is the function to use to check if the pointer is valid or not.

Well it should be, in theory, but as the issue kind of points out, it doesn't actually work in 3.x.

Until there is a proper fix, on a practical level, people will be interested in workarounds such as WeakRef or manually looking up the object ID to see whether it is still valid.

**Comment by Zylann** [comment]
From what I understood of the underlying issue, even WeakRef was subject to this because the problem lies into how ObjectID is looked up? (i.e a given ObjectID can be later re-used)

**Comment by lawnjelly** [comment]
Ah no, hence the misunderstanding. As I remember it, the ObjectID is not reused, that can be seen to change, it was the object pointer / reference (from a pool maybe?) that seems to be reused. This is how I debugged what was going on, from memory. And that is why WeakRef worked when `is_instance_valid()` does not.

**Comment by akien-mga** [comment]
> > Either way, `is_instance_valid()` is the function to use to check if the pointer is valid or not.
> 
> Well it should be, in theory, but as the issue kind of points out, it doesn't actually work in 3.x.

AFAIK this issue has been fixed in `3.3` (actually in `3.2.2`, then with further changes in `3.2.3` which made debug and release behavior inconsistent, fixed in `3.3`). Using `is_instance_valid()` should work. If it doesn't, this should be reported in a new issue.

**Comment by lawnjelly** [comment]
Ah maybe @remram44 is using an old version of the engine, sorry for confusion. :+1:

**Comment by remram44** [comment]
The "old version" I use is the latest stable, 3.3. The problem is fixed in 4.0, but in 3.3 it is possible for a non-weakref pointer to suddenly point to a different object, which is what this issue is about.

**Comment by Sebjugate** [comment]
@aikin

> AFAIK this issue has been fixed in `3.3` (actually in `3.2.2`, then with further changes in `3.2.3` which made debug and release behavior inconsistent, fixed in `3.3`). Using `is_instance_valid()` should work. If it doesn't, this should be reported in a new issue.

Unfortunately, this particular behavior (freed references pointing to random other objects)  is NOT fixed in 3.3.  If you're referring to the dangling pointer fix (#38119 by @RandomShaper ), I'm not entirely sure I understand what it was fixing, but it did not fix this issue.

This seems like a fairly critical issue, and does not break or provide any error or warning in debug, but exhibits this unexpected behavior and problematic behavior in release (references pointing to random other objects).

I can provide a sample project if needed.

From what I've read of @reduz comments elsewhere, this may be impossible to fix in 3.x. If so, is there a document that explains what the best, safest way to store a reference to a node is then?

Thanks everyone!

**Comment by RandomShaper** [comment]
It's fixed in 3.x for run in the editor. About extending it to release builds (and debug without debugger attached), I opened this: https://github.com/godotengine/godot-proposals/issues/1589

If you have a case where it fails running from the editor, please report.

**Comment by remram44** [comment]
I thought the fix was to use `WeakRef`?

**Comment by jaykyburz** [comment]
This issue killed Godot for me, as well as reduz attitude that its not big deal. And then to double down and allow a release to go out where debug is different to release just shows the folks running the show here really have no idea what its like to ship and maintain a real game.  You just can't trust the foundations of this engine, no matter how nice the graphics are. You can't trust the people that guide its development to make good decisions.  

You would have to be crazy to try and build a business on Gotdot. 

Can you imagine shipping something on Steam and then getting reports of random crashes in literately any place in your game because object references are reused. Then you spend days and weeks try to reproduce the issue in a debug build so you can try and work out what is going on, but it doesn't happen in debug.

**Comment by Yekez** [comment]
> This issue killed Godot for me, as well as reduz attitude that its no big deal. And then to double down and allow a release to go out where debug is different to release just shows the folks running the show here really have no idea what its like to ship and maintain a real game. You just can't trust the foundations of this engine, no matter how nice the graphics are. You can't trust the people that guide its development to make good decisions.
> 
> You would have to be crazy to try and build a business on Gotdot.
> 
> Can you imagine shipping something on Steam and then getting reports of random crashes in literately any place in your game because object references are reused. Then you spend days and weeks try to reproduce the issue in a debug build so you can try and work out what is going on, but it doesn't happen in debug.



> This issue killed Godot for me, as well as reduz attitude that its no big deal. And then to double down and allow a release to go out where debug is different to release just shows the folks running the show here really have no idea what its like to ship and maintain a real game. You just can't trust the foundations of this engine, no matter how nice the graphics are. You can't trust the people that guide its development to make good decisions.
> 
> You would have to be crazy to try and build a business on Gotdot.
> 
> Can you imagine shipping something on Steam and then getting reports of random crashes in literately any place in your game because object references are reused. Then you spend days and weeks try to reproduce the issue in a debug build so you can try and work out what is going on, but it doesn't happen in debug.

Totally. This is me getting owned in playstore. Dev attitude is incredible on this one like it is some non important issue. But you never know when it would crash your game without throwing anything, then days and nights to find it. I loved godot man.. I really did, but my game dev career ended before it began.

**Comment by Sebjugate** [comment]
@RandomShaper I think the big issue is that it works _silently_ in Debug, but causes game-breaking bugs in Release. This difference of behavior combined with the fact that Debug does not give any warning or error makes for a really frustrating situation. I guess that means I wholeheartedly support your proposal to make release behavior the same as debug! Thanks for that proposal.

@remram44 I sounds like that might be a work-around. I haven't found any real official documentation that states this is how you should be storing references to nodes. Instead, everywhere I've seen (official and unofficial) promotes storing a direct reference like `var reference = get_node('MyNode')`. 

@jaykyburz @Yekez I still love Godot, but I am only a hobbyist. I can imagine how frustrating this issue would be when trying to release a production game. Especially if you didn't know of this footgun from the beginning.

Sorry to open this can of worms again everyone! I'll go over and support @RandomShaper's proposal. 
IMO, since this particular issue is only half solved (for debug only) it should not be closed.

**Comment by Calinou** [comment]
> IMO, since this particular issue is only half solved (for debug only) it should not be closed.

There is https://github.com/godotengine/godot/issues/50968 now.

**Comment by Sebjugate** [comment]
@Calinou I don't know the technical details, but that sounds like a different issue. Is it due to the same functionality in the backend?

**Comment by RandomShaper** [comment]
> I think the big issue is that it works silently in Debug, but causes game-breaking bugs in Release. This difference of behavior combined with the fact that Debug does not give any warning or error makes for a really frustrating situation.
@Sebjugate, again, in debug, as long as you run from the editor, so the debugger is attached, you should be getting an error message whenever you try to do anything on a variable that has a deleted object assigned. If you find any situation under which that's not true, that's a bug. Please report and it'll be fixed.

Regarding how to store references to nodes, the standard practice is just that. `WeakRef` won't help there.

**Comment by Sebjugate** [comment]
@RandomShaper maybe I am misunderstanding but I do have a situation where I am not getting any error or warning in debug, but a game-breaking bug in release:

In debug, `is_instance_valid(deleted_object)` is always and forever `false` (which is expected, and very good!). However, in release `is_instance_valid(deleted_object)` can quickly and unexpectedly become true _and_ `deleted_object` is now referencing an entirely different node. This causes strange and difficult to find bug because debug works as expected, and does _not_ give any warning when merely calling `is_instance_valid()`.

I apologize if I'm not being clear. I can upload my sample project later this afternoon if that would help.

**Comment by RandomShaper** [comment]
`is_instance_valid()` won't give you an error in debug because calling it on a freed object is a valid use case.  However, in debug it won't have false positives, whereas in release that's possible. That would be solved indeed if release had the same kind of protections as debug, or at least a relevant subset of them.

**Comment by Sebjugate** [comment]
@RandomShaper I understand. It doesn't make sense for debug to throw a warning every time you use `is_instance_valid()` on a freed object. It's just particularly frustrating because (based on my apparently flawed understanding), I used `is_instance_valid()` to guard against this very problem, but instead it has hidden the problem and made it _almost_ impossible to debug.

Regardless, I think we both agree that your proposal to make release work the same as debug should be accepted ASAP. Is there anything we can do as users to help that happen?

**Comment by RandomShaper** [comment]
I believe the best thing is to add your real use case issues to the proposal, but I think you've already done it. If anything, this last specific part about `is_instance_valid()` being a red herring seems like something very relevant to add there.

**Comment by S0yKaf** [comment]
So what is the current best way to handle checking if an object is deleted if I can't reliably trust `is_instance_valid()` in release? it caused a bunch of really nasty and sometimes quite humorous bugs in my release build and I'm now trying to phase all references to it in my code, but I'd like to know if anyone has found an elegant solution to remediate the problem? I even considered building my own version of Godot which would implement whatever behavior debug has for this in the release builds.

**Comment by Sebjugate** [comment]
I have stored the node's instance ID (retrieved from `node.get_instance_id()`) and fetched it with `instance_from_id()`. From my understanding this should be safe, and is more or less what `weakref()` does, except more explicitly (in my opinion). Interestingly, I've found this to be fairly performant. Still about 50% of the performance of storing a direct reference, but `weakref()` was a mere 25%.

I would be interested to what a developer or someone more intimate with the inner workings of Godot has to say though.
